nutils/_util.py

```diff
@@ -766,4 +768,52 @@ def cli(f, *, argv=None):
     return f(**kwargs)
 
 
+def merge_index_map(nin: int, merge_sets: Iterable[Sequence[int]]) -> Tuple[numpy.ndarray, int]:
+    '''Returns an index map relating ``nin`` unmerged elements to ``nout`` merged elements.
+
+    The index map, an array of length ``nin``, satisfies the following conditions:
+
+    *   For every merge set in ``merge_sets``: for every pair of indices ``i``
+        and ``j`` in a merge set, ``index_map[i] == index_map[j]`` is true.
+
+        In code, the following is true:
+
+            all(index_map[i] == index_map[j] for i, *js in merge_sets for j in js)
+
+    *   Selecting the first occurences of indices in ``index_map`` gives the
+        sequence ``range(nout)``.
+
+    Args
+    ----
+    nin : :class:`int`
+        The number of elements before merging.
+    merge_sets : iterable of sequences of at least one :class:`int`
+        An iterable of merge sets, where each merge set lists the indices of
+        input elements that should be merged. Every merge set should have at
+        least one index.
+
+    Returns
+    -------
+    index_map : :class:`numpy.ndarray`
+        Index map with satisfying the above conditions.
+    nout : :class:`int`
+        The number of output indices.
+    '''
+
+    index_map = numpy.arange(nin)
+    def resolve(index):
+        parent = index_map[index]
+        while index != parent:
+            index = parent
+            parent = index_map[index]
+        return index
+    for merge_set in merge_sets:
+        resolved = list(map(resolve, merge_set))
+        index_map[resolved] = min(resolved)
+    new_indices = itertools.count()
+    for iin, ptr in enumerate(index_map):
+        index_map[iin] = next(new_indices) if iin == ptr else index_map[ptr]
+    return index_map, next(new_indices)
```

Comment:
Less elegant, but could `numpy.unique` be actually faster? together with the previous comment that would amount to:

```python
index_map = numpy.arange(nin)
for merge_set in merge_sets:
    resolved = list(merge_set)
    resolved.extend(index_map[resolved])
    index_map[resolved] = min(resolved)
unique, index = numpy.unique(index_map, return_inverse=True)
return index, len(unique)
```

Reply:
This also doesn't apply anymore then.